src/main/java/org/apache/jasper/servlet/JasperLoader.java

```diff
@@ -130,6 +137,32 @@ public JasperLoader(URL[] urls, ClassLoader parent,
         return findClass(name);
     }
 
+    @Override
+    protected Class<?> findClass(final String name) throws ClassNotFoundException {
+        try {
+            return AccessController.doPrivileged(
+                    new PrivilegedExceptionAction<Class>() {
+                        public Class run() throws ClassNotFoundException {
+                            String path = name.replace('.', '/').concat(".class");
+                            URL res = findResource(path);
+                            if (res != null) {
+                                try (InputStream stream = res.openStream()) {
+                                    int length = stream.available();
+                                    byte[] bytes = new byte[length];
+                                    stream.read(bytes);
```

Comment:
I don't think this is guaranteed to fill the buffer (at least not according to the general InputStream contract). You need to check the return value and read in a loop.